### PR TITLE
Fix: refactor HTTP error codes POST /login for Users service

### LINKED ISSUE #624: Fix: refactor HTTP error codes POST /login for Users service
## Description
As a GSoC student,
I need the correct HTTP error response code to be returned accordingly by Mentorship System API in relation to POST /login,
so that I can represent the error message accordingly to both Mentorship System and Bridge In Tech users.

## Mocks
NA

### Update [Required]
- [ ] 401: Unauthorized needs to be returned instead of 404: Not Found if the username and/or password is/are wrong

### Enhancement to Update [Optional]
NA

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

## Estimation
1 hour

### PR DESCRIPTION
**Pull request by PareenKarpe**
### Description

Fixes #624 

### Type of Change:

- Code
- Quality Assurance
- User Interface
- Outreach
- Documentation

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Tried to login with an incorrect password and/or username. The response type of POST login user service was 401.

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged
- [ ] Update Postman API at /docs folder
- [ ] Update Swagger documentation and the exported file at /docs folder
- [ ] Update requirements.txt

**Code/Quality Assurance Only**

- [x] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Review by satya7289 — COMMENTED** [review]
Look great. 
One more thing need to do, write one test case (here - `tests/users/test_api_login.py`) to validate that it is responding to correct error code.

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@PareenKarpe this looks great! 👏 However to make sure this is tested from now on, could you please include a unit test, just like @satya7289 mentioned 🙏 .
Let me know if you have any doubt that we could help.

**Comment by PareenKarpe** [comment]
> @PareenKarpe this looks great! 👏 However to make sure this is tested from now on, could you please include a unit test, just like @satya7289 mentioned 🙏 .
> Let me know if you have any doubt that we could help.

@isabelcosta Sure I will update the test cases for it

File: tests/users/test_api_login.py [hunk 1]
```diff
@@ -43,6 +45,54 @@ def setUp(self):
...
+            self.assertIsNone(response.json.get("access_expiry"))
+            self.assertIsNone(response.json.get("refresh_token"))
+            self.assertIsNone(response.json.get("refresh_expiry"))
+
+            self.assertEqual(1, len(response.json))
+            self.assertEqual(
+                messages.WRONG_USERNAME_OR_PASSWORD, response.json
+            )
+
+            self.assertEqual(401, response.status_code)
```
Comment (satya7289) [inline tests/users/test_api_login.py]: Here, instead of 401, use `HTTPStatus.UNAUTHORIZED`.

File: tests/users/test_api_login.py [hunk 2]
```diff
@@ -43,6 +45,54 @@ def setUp(self):
...
```
Comment (satya7289) [inline tests/users/test_api_login.py]: Here, instead of 401, use `HTTPStatus.UNAUTHORIZED`.

**Review by satya7289 — COMMENTED** [review]
Update the swagger docs also.(`@users_ns.response(HTTPStatus.NOT_FOUND, "%s" % messages.WRONG_USERNAME_OR_PASSWORD)`)

File: tests/users/test_api_login.py (hunk 2 above)
Comment (PareenKarpe) [reply tests/users/test_api_login.py]: Sorry I missed that. I will update it.

File: tests/users/test_api_login.py [hunk 3]
```diff
@@ -43,6 +46,54 @@ def setUp(self):
         db.session.add(self.first_user)
         db.session.add(self.second_user)
         db.session.commit()
+    def test_user_login_invalid_credentials(self):
```
Comment (isabelcosta) [inline tests/users/test_api_login.py]: add a blank line below please

File: tests/users/test_api_login.py [hunk 4]
```diff
@@ -43,6 +46,54 @@ def setUp(self):
...
+            self.assertIsNone(response.json.get("refresh_token"))
+            self.assertIsNone(response.json.get("refresh_expiry"))
+
+            self.assertEqual(1, len(response.json))
+            self.assertEqual(
+                messages.WRONG_USERNAME_OR_PASSWORD, response.json
+            )
+
+            self.assertEqual(HTTPStatus.UNAUTHORIZED, response.status_code)
+    def test_user_login_invalid_credentials(self):
```
Comment (isabelcosta) [inline tests/users/test_api_login.py:50]: please add a blank line above, in line 72. It for readability concerns. Unfortunately, we don't have a lint checker as part of the checks, yet, so I am just being a human linter 🙈

Comment (isabelcosta) [reply tests/users/test_api_login.py:50]: oh wait, isn't this test the exact same as the above one? If so, can you please remove it?

**Review by isabelcosta — CHANGES_REQUESTED** [review]
I think you might have a duplicated test here, if so, can you please remove it?
This is almost done ;) thank you for your contribution @PareenKarpe 👏

File: tests/users/test_api_login.py (hunk 4 above)
Comment (PareenKarpe) [reply tests/users/test_api_login.py:50]: Oh yes, you are correct. Thanks for checking. I was planning to add separate test cases and I missed it. I am removing the extra code.

**Comment by PareenKarpe** [comment]
> Update the swagger docs also.(`@users_ns.response(HTTPStatus.NOT_FOUND, "%s" % messages.WRONG_USERNAME_OR_PASSWORD)`)

@satya7289 For the swagger docs, I have updated the code in user.py. Am I missing any other change related to it?

**Comment by PareenKarpe** [comment]
> Update the swagger docs also.(`@users_ns.response(HTTPStatus.NOT_FOUND, "%s" % messages.WRONG_USERNAME_OR_PASSWORD)`)

For swagger docs, I made changes in user.py. Am I missing any other change related to it?

File: tests/users/test_api_login.py (hunk 4 above)
Comment (PareenKarpe) [reply tests/users/test_api_login.py:50]: I have removed the duplicate test case

**Comment by PareenKarpe** [comment]
> I think you might have a duplicated test here, if so, can you please remove it?
> This is almost done ;) thank you for your contribution @PareenKarpe 👏

@isabelcosta Thank you for all the suggestions and corrections. It helped me 👍 . I have removed the duplicate test case

File: tests/users/test_api_login.py [hunk 5]
```diff
@@ -43,6 +46,29 @@ def setUp(self):
         db.session.add(self.first_user)
         db.session.add(self.second_user)
         db.session.commit()
+    def test_user_login_invalid_credentials(self):
```
Comment (isabelcosta) [inline tests/users/test_api_login.py:50]:
```suggestion

    def test_user_login_invalid_credentials(self):
```

Comment (isabelcosta) [reply tests/users/test_api_login.py:50]: @PareenKarpe can you please do this? I think it would make the code more readable :)

Comment (PareenKarpe) [reply tests/users/test_api_login.py:50]: @isabelcosta Sure. I made those changes.

**Review by isabelcosta — APPROVED** [review]
@PareenKarpe great work! I'm approving this :)
Thank you so much for your patience and contribution! 🎉

**Comment by PareenKarpe** [comment]
> @PareenKarpe great work! I'm approving this :)
> Thank you so much for your patience and contribution! 🎉

Thank you 👍

**Review by foongminwong — APPROVED** [review]
The changes made in this PR were tested locally. Following are the results:

1. Code review - Done

2. All possible responses were tested as below:

   * _Before PR 647_  
    _Screenshot/gif/url_:  

    before-pr-647

    _Expected Result_:  The HTTP error response code to be returned accordingly by Mentorship System API in relation to POST /login should return 404.
    _Actual Result_: Same as expected.

   * _After PR 647_  
    _Screenshot/gif/url_:  

     after-pr-647

    _Expected Result_:  The HTTP error response code to be returned accordingly by Mentorship System API in relation to POST /login should return 401.
    _Actual Result_:  Same as expected.
    
3. Additional testcases covered: N/A

4. Additional Comments: N/A

5. Status of PR Changed to: N/A

6. OS: Wndows 10

**Comment by isabelcosta** [comment]
Thank you so much for testing and performing code review @foongminwong 🤗

**Pull request merged**